File: 3975 - Escape from enemy territory/solutions/b2.java
import java.util.*;

public class b2 {

  static Scanner in = new Scanner(System.in);

  double jdist2;
  double[] x,y;
  int[] npeng, mjumps;

  int adj[][], cap[][];
  boolean[] used;
  
  int dfs(int n, int t, int c) {
    used[n]=true;
    if(n==t) {
      return c;
    }

    for(int i=0; i<cap.length; i++) if(!used[i] && cap[n][i]>0) {
      int pathc=dfs(i, t, Math.min(cap[n][i], c));
      if(pathc>0) {
        cap[n][i]-=pathc;
        cap[i][n]+=pathc;
        return pathc;
      }
    }
    return 0;
  }

  int maxFlow(int s, int t, int[][] _cap) {
    int flow=0;
    int pathc=0;
    int N=_cap.length;
    cap=new int[N][N];
    for(int i=0; i<N; i++) for(int j=0; j<N; j++) cap[i][j]=_cap[i][j]; // kut java
    do {
      used=new boolean[cap.length];
      pathc=dfs(s, t, 1234567890);
      flow+=pathc;
    } while(pathc!=0);
    return flow;
  }

  boolean isEdge(int m, int n) {
    double dx=x[m]-x[n], dy=y[m]-y[n];
    return dx*dx+dy*dy<jdist2;
  }

  void go() {
    int N=in.nextInt(); double jdist=in.nextDouble();
    jdist2=jdist*jdist;
    x=new double[N]; y=new double[N];
    npeng=new int[N]; mjumps=new int[N];
    for(int i=0; i<N; i++) {
      x[i]=in.nextDouble(); y[i]=in.nextDouble();
      npeng[i]=in.nextInt(); mjumps[i]=in.nextInt();
    }

    int[][] cap=new int[2*N+1][2*N+1];

    int totpeng=0;
    for(int i=0; i<N; i++) {
      totpeng+=npeng[i];
    // edges from source to in vertex ice cube (cap is num penguins)
    //  adj[2*N].PB(i); adj[i].PB(2*N);
      cap[2*N][i]=npeng[i];
    // edges from in vertex to out vertex (cap is max jumps)
    //  adj[i].PB(i+N); adj[i+N].PB(i);
      cap[i][i+N]=mjumps[i];
      for(int j=0; j<N; j++) if(i!=j && isEdge(i,j)) {
      // edges from out vertex to in vertex if jumpable (cap is infinite)
      //  adj[i+N].PB(j); adj[j].PB(i+N);
        cap[i+N][j]=1234567890;
      }
    }

    boolean first=true;
    for(int i=0; i<N; i++) if(maxFlow(2*N, i, cap)==totpeng) {
      if(!first) System.out.print(" ");
      first=false;
      System.out.print(i);
    }
    if(first) System.out.println("-1");
    else System.out.print("\n");
  }

  public static void main(String[] args) throws Exception {
    int nruns = in.nextInt();
    while (nruns-- > 0) {
      new b2().go();
    }
  }

}

File: 3975 - Escape from enemy territory/solutions/b2.cpp
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <cctype>

#include <algorithm>
#include <bitset>
#include <complex>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

#define PB push_back
#define ALL(x) (x).begin(),(x).end()
#define SZ(x) ((int)(x).size())
#define REP(i, n) for(int i=0; i<n; ++i)
#define REPD(i, n) for(int i=(n)-1; i>=0; --i)
#define FOR(i, b, e) for(typeof(e) i=b; i!=e; ++i)

using namespace std;

typedef long long LL;
typedef vector<int> VI;
typedef vector<VI> VVI;
typedef vector<string> VS;
typedef istringstream ISS;

int N;
VI x, y, npeng, mjumps;

double jdist;

const int INF=10000000;

VVI adj;

int find(int s, int t, VVI &cap) {
  int N=SZ(cap);

  VI prevnode(N,-1);
  vector<int> flow(N, 0);
  deque<int> q;
  q.PB(s);
  flow[s]=INF;

  while(!q.empty()) {
    int node=q.front();
    q.pop_front();

    if(node==t) break;

    REP(i, SZ(adj[node])) {
//    REP(i, N) {
      int next=adj[node][i];
      if(flow[next] || cap[node][next]==0) continue;
      flow[next]=min(flow[node], cap[node][next]);
      prevnode[next]=node;
      q.PB(next);
    }
  }
  //if(!visited[t]) return 0;

  int fl=flow[t];
  while(prevnode[t]!=-1) {
    cap[t][prevnode[t]]+=fl;
    cap[prevnode[t]][t]-=fl;
    t=prevnode[t];
  }
  return fl;
}

int maxflow(int s, int t, VVI cap) {
  int flow=0;
  int pathcap=0;
  do {
    pathcap=find(s, t, cap);
    flow+=pathcap;
  } while(pathcap);
  return flow;
}

bool isedge(int i ,int j) {
  double dx=x[i]-x[j], dy=y[i]-y[j];
  return dx*dx+dy*dy<jdist*jdist+1e-9;
}

void go() {
  scanf("%d %lf", &N, &jdist);
  x.resize(N); y.resize(N);
  npeng.resize(N); mjumps.resize(N);

  REP(i, N) scanf("%d %d %d %d", &x[i], &y[i], &npeng[i], &mjumps[i]);

  adj=VVI(2*N+1);
  VVI cap(2*N+1, VI(2*N+1, 0));

  int totpeng=0;
  REP(i, N) {
    totpeng+=npeng[i];
    // edges from source to in vertex ice cube (cap is num penguins)
    adj[2*N].PB(i); adj[i].PB(2*N);
    cap[2*N][i]=npeng[i];
    // edges from in vertex to out vertex (cap is max jumps)
    adj[i].PB(i+N); adj[i+N].PB(i);
    cap[i][i+N]=mjumps[i];
    REP(j, N) if(i!=j && isedge(i,j)) {
      // edges from out vertex to in vertex if jumpable (cap is infinite)
      adj[i+N].PB(j); adj[j].PB(i+N);
      cap[i+N][j]=INF;
    }
  }

  VI res;
  REP(i, N) if(maxflow(2*N, i, cap)==totpeng) res.PB(i);
  if(res.empty()) printf("-1\n");
  else {
    printf("%d", res[0]);
    FOR(i, 1, SZ(res)) printf(" %d", res[i]);
    printf("\n");
  }
}

int main() {
  int nruns;

  scanf("%d", &nruns);
  while(nruns--) go();

  return 0;
}
